### PR TITLE
Fix setting reset header to gauge histogram in seriesToChunkEncoder

### PR DESCRIPTION
**Pull request by krajorama**

### CODE REVIEW

File: storage/series_test.go [hunk 1]
```diff
@@ -145,7 +144,7 @@ func TestHistogramSeriesToChunks(t *testing.T) {
 	}
 	// Appendable to h1.
 	h2 := &histogram.Histogram{
-		Count:         12,
+		Count:         10,
```
Comment (beorn7) [inline storage/series_test.go:147]: Why this change? The "correct" count is indeed 12.

File: storage/series_test.go [hunk 2]
```diff
@@ -127,13 +127,12 @@ func TestChunkSeriesSetToSeriesSet(t *testing.T) {
...
 	samples              []tsdbutil.Sample
-	expectedChunks       int
-	expectedCounterReset bool
+	expectedCounterReset []chunkenc.CounterResetHeader
 }
 
 func TestHistogramSeriesToChunks(t *testing.T) {
 	h1 := &histogram.Histogram{
-		Count:         3,
+		Count:         5,
```
Comment (beorn7) [inline storage/series_test.go:135]: The correct count is actually 7. (Need to include the zero count.)

File: storage/series_test.go [hunk 3]
```diff
@@ -208,6 +207,61 @@ func TestHistogramSeriesToChunks(t *testing.T) {
 		PositiveBuckets: []float64{2, 2, 7, 2},
 	}
 
+	// Gauge histogram
```
Comment (beorn7) [inline storage/series_test.go:210]: Nit: We usually end comments with a period even if they aren't strictly whole sentences.

File: storage/series_test.go [hunk 4]
```diff
@@ -208,6 +207,61 @@ func TestHistogramSeriesToChunks(t *testing.T) {
...
+		Sum:              100,
+		Schema:           0,
+		PositiveSpans: []histogram.Span{
+			{Offset: 0, Length: 2},
+			{Offset: 1, Length: 2},
+		},
+		PositiveBuckets: []int64{2, 1, -2, 3}, // Abs: 2, 3, 1, 4
+	}
+
+	// Float gauge histogram
```
Comment (beorn7) [inline storage/series_test.go:237]: End with period as above.

File: storage/series_test.go [hunk 5]
```diff
@@ -208,6 +207,61 @@ func TestHistogramSeriesToChunks(t *testing.T) {
 		PositiveBuckets: []float64{2, 2, 7, 2},
 	}
 
+	// Gauge histogram
+	gh1 := &histogram.Histogram{
+		CounterResetHint: histogram.GaugeType,
+		Count:            5,
```
Comment (beorn7) [inline storage/series_test.go:213]: Following the theme of "correct counts": This should be 7.

File: storage/series_test.go [hunk 6]
```diff
@@ -208,6 +207,61 @@ func TestHistogramSeriesToChunks(t *testing.T) {
...
+		Sum:              100,
+		Schema:           0,
+		PositiveSpans: []histogram.Span{
+			{Offset: 0, Length: 2},
+		},
+		PositiveBuckets: []int64{2, 1}, // Abs: 2, 3
+	}
+	gh2 := &histogram.Histogram{
+		CounterResetHint: histogram.GaugeType,
+		Count:            10,
```
Comment (beorn7) [inline storage/series_test.go:225]: And 12 here.

File: storage/series_test.go [hunk 7]
```diff
@@ -208,6 +207,61 @@ func TestHistogramSeriesToChunks(t *testing.T) {
...
+			{Offset: 0, Length: 2},
+			{Offset: 1, Length: 2},
+		},
+		PositiveBuckets: []int64{2, 1, -2, 3}, // Abs: 2, 3, 1, 4
+	}
+
+	// Float gauge histogram
+	gfh1 := &histogram.FloatHistogram{
+		CounterResetHint: histogram.GaugeType,
+		Count:            4,
```
Comment (beorn7) [inline storage/series_test.go:240]: 6.

File: storage/series_test.go [hunk 8]
```diff
@@ -127,13 +127,12 @@ func TestChunkSeriesSetToSeriesSet(t *testing.T) {
 
 type histogramTest struct {
 	samples              []tsdbutil.Sample
-	expectedChunks       int
-	expectedCounterReset bool
+	expectedCounterReset []chunkenc.CounterResetHeader
```
Comment (beorn7) [inline storage/series_test.go:130]: Maybe better call in in the plural, i.e. `expectedCounterResets`

**Review by beorn7 — CHANGES_REQUESTED** [review]
This looks right to me, but @codesome is probably a million times more qualified to judge.

My comments are just the usual punctuation nits etc.

File: storage/series_test.go (hunk 1 above)
Comment (krajorama) [reply storage/series_test.go:147]: I was tired, forgot about the zero count, thank you for the review, fixing the findings.

File: storage/series_test.go [hunk 9]
```diff
@@ -208,6 +207,61 @@ func TestHistogramSeriesToChunks(t *testing.T) {
...
+		ZeroCount:        2,
+		ZeroThreshold:    0.001,
+		Sum:              100,
+		Schema:           0,
+		PositiveSpans: []histogram.Span{
+			{Offset: 0, Length: 2},
+		},
+		PositiveBuckets: []float64{3, 1},
+	}
+	// Appendable to fh1.
```
Comment (codesome) [inline storage/series_test.go:250]: I guess this comment needs to be removed?

File: storage/series_test.go [hunk 10]
```diff
@@ -126,14 +126,13 @@ func TestChunkSeriesSetToSeriesSet(t *testing.T) {
 }
 
 type histogramTest struct {
-	samples              []tsdbutil.Sample
-	expectedChunks       int
-	expectedCounterReset bool
+	samples               []tsdbutil.Sample
+	expectedCounterResets []chunkenc.CounterResetHeader
```
Comment (codesome) [inline storage/series_test.go:130]: Or maybe `expCounterResetHeaders`
```suggestion
	expectedCounterResetHeaders []chunkenc.CounterResetHeader
```

**Review by codesome — APPROVED** [review]
LGTM with these tiny nits. Thanks!

**Review by codesome — CHANGES_REQUESTED** [review]
Undoing my previous approval because the DCO is failing. Can you also fix it, please?

File: storage/series_test.go [hunk 11]
```diff
@@ -322,13 +411,24 @@ func TestHistogramSeriesToChunks(t *testing.T) {
 
 func testHistogramsSeriesToChunks(t *testing.T, test histogramTest) {
 	lbs := labels.FromStrings("__name__", "up", "instance", "localhost:8080")
-	series := NewListSeries(lbs, test.samples)
+	copiedSamples := []tsdbutil.Sample{}
```
Comment (krajorama) [inline storage/series_test.go:414]: @codesome I had to add this copy for gauge histograms as the histogram was modified in place (spans and buckets, nothing else). Is that expected?

Comment (codesome) [reply storage/series_test.go:414]: Copied in `NewListSeries` or somewhere else?

Comment (codesome) [reply storage/series_test.go:414]: I think the modification is happening in the `seriesToChunkEncoder` when there is recoding. The recoding also happens in TSDB append, and if it mutating the histogram, it is likely breaking the interface as described [here](https://github.com/prometheus/prometheus/blob/7c2de14b0bd74303c2ca6f932b71d4585a29ca75/storage/interface.go#L282-L284). I will open a separate conversation on that.

**Review by codesome — APPROVED** [review]

File: storage/series_test.go (hunk 11 above)
Comment (krajorama) [reply storage/series_test.go:414]: This is coming from https://github.com/prometheus/prometheus/blob/7c2de14b0bd74303c2ca6f932b71d4585a29ca75/storage/series.go#L482  Which has "Recoding" in the name, and calls RecordHistogram which does the change in place. So it's kind of understandable...
